### PR TITLE
Adding gradient for Concat

### PR DESCRIPTION
**Pull request by jonathankochems**
Hi,

This pull request adds the gradient for Concat ops. Any feedback or suggestions are welcome.

Jonathan

### CODE REVIEW

**Comment by blackgnezdo** [comment]
Jenkins test this please.

**Comment by blackgnezdo** [comment]
Thanks a lot for the PR and the thorough test coverage!

@fkm3, could you please do the review?

**Comment by blackgnezdo** [comment]
Jenkins test this please.

**Comment by fkm3** [comment]
Yeah, I'll take a look tonight.

**Comment by fkm3** [comment]
Jenkins test this please.

File: tensorflow-ops/src/TensorFlow/Gradient.hs [hunk 1]
```diff
@@ -443,6 +444,36 @@ opGrad "Neg" _ [_] [dz] = [Just $ negate $ expr dz]
...
+ | length x == 1 = Nothing : [Just $ expr dy]
+ | otherwise     = Nothing : map Just (dx `reshapeZip` s)
+   where x  :: [Tensor Build a]
+         x  = map toT $ tail _ix
+         _i = toT $ head _ix
+         i  = reshape _i one
+         m  = length x
+         s  :: [Tensor Build Int32]
+         s  = map shape x
+         di :: Tensor Build Int32
```
Comment (jonathankochems) [inline tensorflow-ops/src/TensorFlow/Gradient.hs]: Makes sense. I've renamed the sizes along the concat dimension to ki. I hope that is clearer.

File: tensorflow-ops/src/TensorFlow/Gradient.hs [hunk 2]
```diff
@@ -443,6 +444,36 @@ opGrad "Neg" _ [_] [dz] = [Just $ negate $ expr dz]
...
+         di = CoreOps.concat (scalar 0) $ map (\t -> CoreOps.slice t i one) s
```
Comment (jonathankochems) [inline tensorflow-ops/src/TensorFlow/Gradient.hs]: Thanks. I'll adopt that.

File: tensorflow-ops/src/TensorFlow/Gradient.hs [hunk 3]
```diff
@@ -443,6 +444,36 @@ opGrad "Neg" _ [_] [dz] = [Just $ negate $ expr dz]
...
+         dx = CoreOps.splitV (fromIntegral m) dy di _i
+         reshapeZip = zipWith reshape
+         one = constant (Shape [1 :: Int64]) [1 :: Int32]
```
Comment (jonathankochems) [inline tensorflow-ops/src/TensorFlow/Gradient.hs]: Nice, I've worked that in.

File: tensorflow-ops/tests/GradientTest.hs [hunk 4]
```diff
@@ -164,6 +167,117 @@ testMaxGradient = testCase "testMaxGradient" $ do
...
+    dinputs <- TF.runSession $ do
+        inputTensors <- forM (inputs `zip` shapes) $ \(input,shape) ->
+                          TF.render $ TF.constant (TF.Shape shape) input
+        inputGradTensors <- forM (inputGrads `zip` shapes) $ \(inputGrad, shape) ->
+                               TF.render $ TF.constant (TF.Shape shape) inputGrad
+        inputGradTensor <- TF.render $ TF.concat (TF.scalar 1) inputGradTensors
+        inputTensor <- TF.render $ TF.concat (TF.scalar 1) inputTensors
+        output <- TF.render $ inputTensor `TF.mul` inputGradTensor
+        TF.gradients output inputTensors >>= TF.run
+    (V.fromList <$> inputGrads) @=? dinputs
```
Comment (jonathankochems) [inline tensorflow-ops/tests/GradientTest.hs]: Yes, fair enough. I've refactored out the common parts and also expanded the tests to cover negative indices.

File: tensorflow-ops/src/TensorFlow/Gradient.hs [hunk 5]
```diff
@@ -443,6 +444,36 @@ opGrad "Neg" _ [_] [dz] = [Just $ negate $ expr dz]
...
+--  The incoming gradient from backpropagation is
+--   simply forwarded split across input tensors.
+--   Forwarded gradients have shapes s = [s1, ..., sm].
+opGrad "Concat" _ _ix [dy]
+ | length x == 1 = Nothing : [Just $ expr dy]
+ | otherwise     = Nothing : map Just (dx `reshapeZip` s)
+   where x  :: [Tensor Build a]
+         x  = map toT $ tail _ix
+         _i = toT $ head _ix
+         i  = reshape _i one
```
Comment (jonathankochems) [inline tensorflow-ops/src/TensorFlow/Gradient.hs]: Thanks for spotting this. Indeed splitV does not handle negative indices, so I've adopted your suggestion.

File: tensorflow-ops/src/TensorFlow/Gradient.hs [hunk 6]
```diff
@@ -443,6 +444,36 @@ opGrad "Neg" _ [_] [dz] = [Just $ negate $ expr dz]
...
+--    .           .     .          .        .
+--   xm of shape sm = [d1, ..., di_m, ..., dn]
+--  along dimension i to an output tensor
+--   y  of shape sy = [d1, ..., d, ..., dn]
+--  where d = sum di = sum [di_1,...,di_m]
+--
+--  The incoming gradient from backpropagation is
+--   simply forwarded split across input tensors.
+--   Forwarded gradients have shapes s = [s1, ..., sm].
+opGrad "Concat" _ _ix [dy]
```
Comment (jonathankochems) [inline tensorflow-ops/src/TensorFlow/Gradient.hs]: Ok. I've tried to respect this in my latest changes. Is your whitespace convention documented somewhere?

**Comment by fkm3** [comment]
Jenkins test this please.

File: tensorflow-ops/src/TensorFlow/Gradient.hs (hunk 6 above)
Comment (fkm3) [inline tensorflow-ops/src/TensorFlow/Gradient.hs]: Our internal style guide hasn't been published yet, unfortunately.

**Comment by fkm3** [comment]
Looks good, thanks!
